### PR TITLE
Improve account detection

### PR DESCRIPTION
**Pull request by yabirgb**
This PR changes how accounts get automatically detected:

1. If the account had any transaction or internal transaction as categorized in etherscan it is added
2. if it had only erc20 transfers we iterate over the logs filtering the transactions. If any token is uknown the account is marked as not spammed but if all of them are ignored or marked as spam the account gets marked as spammed

### CODE REVIEW

File: rotkehlchen/externalapis/etherscan.py [hunk 1]
```diff
@@ -474,20 +481,20 @@ def get_token_transaction_hashes(
 
         yield _hashes_tuple_to_list(hashes)
 
-    def has_activity(self, account: ChecksumEvmAddress) -> bool:
+    def has_activity(self, account: ChecksumEvmAddress) -> EtherscanHasChainActivity:
```
Comment (LefterisJP) [inline rotkehlchen/externalapis/etherscan.py:484]: please modify the docstring a bit to mention the different return values

File: rotkehlchen/externalapis/etherscan.py [hunk 2]
```diff
@@ -53,6 +54,12 @@
 log = RotkehlchenLogsAdapter(logger)
 
 
+class EtherscanHasChainActivity(Enum):
```
Comment (LefterisJP) [inline rotkehlchen/externalapis/etherscan.py:57]: short docstring explainign what this is

File: rotkehlchen/chain/evm/transactions.py [hunk 3]
```diff
@@ -447,6 +460,72 @@ def _get_erc20_transfers_for_ranges(
                 queried_ranges=[(start_ts, end_ts)],
             )
 
+    def address_has_been_spammed(self, address: ChecksumEvmAddress) -> bool:
+        """
+        Queries erc20 tranfers for the given address and if it has only spam assets or ignored
+        assets we return True.
+        If any transfer had an unknown asset
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/transactions.py:467]:
```suggestion
        If any transfer had an unknown asset ... then what?
```

File: rotkehlchen/chain/evm/transactions.py [hunk 4]
```diff
@@ -447,6 +460,72 @@ def _get_erc20_transfers_for_ranges(
...
+        """
+        start_ts, end_ts = Timestamp(0), ts_now()
+        checekd_tokens = set()
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/transactions.py:470]:
```suggestion
        checked_tokens = set()
```

File: rotkehlchen/chain/evm/transactions.py [hunk 5]
```diff
@@ -447,6 +460,72 @@ def _get_erc20_transfers_for_ranges(
...
+        log.debug(f'Address detection: querying {self.evm_inquirer.chain_name} ERC20 Transfers for {address} -> {start_ts} - {end_ts}')  # noqa: E501
+        try:
+            for erc20_tx_hashes in self.evm_inquirer.etherscan.get_token_transaction_hashes(
+                account=address,
+                from_ts=start_ts,
+                to_ts=end_ts,
+            ):
+                for tx_hash in erc20_tx_hashes:
+                    tx_hash_bytes = deserialize_evm_tx_hash(tx_hash)
+                    _, raw_receipt_data = self.evm_inquirer.get_transaction_by_hash(tx_hash_bytes)  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/transactions.py:483]: This gets both the transaction hash and the receipt data. You only need the receipts here. So just ask for the receipts with the appropriate function.

File: rotkehlchen/chain/evm/transactions.py [hunk 6]
```diff
@@ -447,6 +460,72 @@ def _get_erc20_transfers_for_ranges(
...
+
+                        # this token is not ignored, is not spam and exists in the database
+                        return False
+
+                    log.debug(f'Address detection: queried {self.evm_inquirer.chain_name} ERC20 Transfers for {address} -> range {start_ts} - {end_ts}')  # noqa: E501
+        except RemoteError as e:
+            self.msg_aggregator.add_error(
+                f'Got error "{e!s}" while querying {self.evm_inquirer.chain_name} '
+                f'token transactions from Etherscan. Transactions not added to the DB '
+                f'address: {address} ',
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/transactions.py:524]: copy pasta

```suggestion
                f'token transactions from Etherscan. address: {address} spam detection failed',
```

File: rotkehlchen/chain/evm/transactions.py [hunk 7]
```diff
@@ -447,6 +460,72 @@ def _get_erc20_transfers_for_ranges(
                 queried_ranges=[(start_ts, end_ts)],
             )
 
+    def address_has_been_spammed(self, address: ChecksumEvmAddress) -> bool:
+        """
+        Queries erc20 tranfers for the given address and if it has only spam assets or ignored
+        assets we return True.
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/transactions.py:466]:
```suggestion
        Queries erc20 tranfers for the given address and if it has only transfer of spam assets or ignored
        assets we return True. Stop at the first valid ERC20 transfer we find to exit as early as possible.
```

File: rotkehlchen/chain/evm/transactions.py [hunk 8]
```diff
@@ -447,6 +460,72 @@ def _get_erc20_transfers_for_ranges(
...
+        checekd_tokens = set()
+        with self.database.conn.read_ctx() as cursor:
+            ignored_assets = self.database.get_ignored_asset_ids(cursor)
+
+        log.debug(f'Address detection: querying {self.evm_inquirer.chain_name} ERC20 Transfers for {address} -> {start_ts} - {end_ts}')  # noqa: E501
+        try:
+            for erc20_tx_hashes in self.evm_inquirer.etherscan.get_token_transaction_hashes(
+                account=address,
+                from_ts=start_ts,
+                to_ts=end_ts,
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/transactions.py:479]: A note. This checks every 10 transactions. A hard-coded choice for all of our etherscan queries. It's the variable `TRANSACTIONS_BATCH_NUM`.

But perhaps in this particular case since we want to exit as early as possible smaller value would have been better. But also perhaps this is overthinking it.

File: rotkehlchen/chain/aggregator.py [hunk 9]
```diff
@@ -1310,18 +1311,28 @@ def check_single_address_activity(
...
+                    etherscan_activity = chain_manager.node_inquirer.etherscan.has_activity(address)  # noqa: E501
+                    if (
+                        (
+                            etherscan_activity == EtherscanHasChainActivity.TOKENS and
+                            chain_manager.transactions.address_has_been_spammed(address=address)
+                        ) or
+                        etherscan_activity == EtherscanHasChainActivity.NONE
+                    ):  # in the case of tokens we check if the transactions were spam tokens
+                        continue
+                    # else we add the chain
```
Comment (LefterisJP) [inline rotkehlchen/chain/aggregator.py:1335]: readability nitpick

```suggestion
                    only_token_spam = (
                        etherscan_activity == EtherscanHasChainActivity.TOKENS and
                        chain_manager.transactions.address_has_been_spammed(address=address)
                    )
                    if only_token_spam or etherscan_activity == EtherscanHasChainActivity.NONE:
                        continue  #  do not add the address for the chain

```

File: rotkehlchen/tests/unit/test_chains_aggregator.py [hunk 10]
```diff
@@ -142,3 +163,62 @@ def test_detect_evm_accounts(blockchain: 'ChainsAggregator') -> None:
...
+
+
+@pytest.mark.vcr(filter_query_parameters=['apikey'])
+@pytest.mark.freeze_time('2023-06-19 05:16:10 GMT')
+@pytest.mark.parametrize('polygon_pos_manager_connect_at_start', [(ALCHEMY_POLYGON_NODE,)])
+@pytest.mark.parametrize('polygon_pos_accounts', [[make_evm_address()]])  # to connect to nodes
+def test_detect_evm_accounts_spam_tx(polygon_pos_manager: 'PolygonPOSManager') -> None:
+    """
+    Test correctly that an account with only erc20 transfers of spam tokens get correctly
+    marked as spammed.
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_chains_aggregator.py:175]:
```suggestion
    Test that an account with only erc20 transfers of spam tokens gets marked as spam
    and does not get detected as a tracked account in the EVM chain.
```

File: rotkehlchen/tests/unit/test_chains_aggregator.py [hunk 11]
```diff
@@ -142,3 +163,62 @@ def test_detect_evm_accounts(blockchain: 'ChainsAggregator') -> None:
...
+    eip155:137/erc20:0x06732174B52743C374445E88C9b01031Bd0FB28f
+    eip155:137/erc20:0xE2Ee00F49464d6B60771dc118A1bb4eb362bd154
+    eip155:137/erc20:0x37CC5F5610d91325c8A8C0eD74d26A01F19e7B51
+
+    first we check that it is marked as NOT spammed by:
+    - ignoring the first asset
+    - adding the second as spam asset
+    - third is unknown so the address is marked as not spammed
+
+    to verify that the transaction has been spamemd all the remaining assets are ignored
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_chains_aggregator.py:193]: What do you mean by the transaction has been spammed?

```suggestion
    to verify that the transaction has been spammed all the remaining assets are ignored
```

This makes no sense in English. Do you mean, marked as spam?

File: rotkehlchen/tests/utils/polygon_pos.py [hunk 12]
```diff
@@ -2,6 +2,7 @@
 from rotkehlchen.fval import FVal
 from rotkehlchen.types import SupportedBlockchain
 
+ALCHEMY_RPC_ENDPOINT = 'https://polygon-mainnet.g.alchemy.com/v2/uNdnI7_6XXc7ayswOxtd_RuTBGojJhIf'
```
Comment (LefterisJP) [inline rotkehlchen/tests/utils/polygon_pos.py:5]: why this change? Where else is this used?

File: rotkehlchen/tests/unit/test_chains_aggregator.py (hunk 11 above)
Comment (yabirgb) [reply rotkehlchen/tests/unit/test_chains_aggregator.py:193]: it is `the address has been spammed`

File: rotkehlchen/chain/aggregator.py [hunk 13]
```diff
@@ -1310,18 +1311,27 @@ def check_single_address_activity(
...
+                else:
+                    etherscan_activity = chain_manager.node_inquirer.etherscan.has_activity(address)  # noqa: E501
+                    only_token_spam = (
+                        etherscan_activity == EtherscanHasChainActivity.TOKENS and
+                        chain_manager.transactions.address_has_been_spammed(address=address)
+                    )
+                    if only_token_spam or etherscan_activity == EtherscanHasChainActivity.NONE:
+                        continue  # do not add the address for the chain
+
+                    # else we add the chain
```
Comment (LefterisJP) [inline rotkehlchen/chain/aggregator.py:1334]: Do not do that. It's obvious that you add the chain later. And if you **really** wanted to add this it should be in line 1339 as part of same line. Not on its own line. Please remove

```suggestion
```

File: rotkehlchen/chain/evm/transactions.py [hunk 14]
```diff
@@ -447,6 +460,71 @@ def _get_erc20_transfers_for_ranges(
...
+            ):
+                for tx_hash in erc20_tx_hashes:
+                    raw_receipt_data = self.evm_inquirer.get_transaction_receipt(deserialize_evm_tx_hash(tx_hash))  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/transactions.py:483]: Hmm I just noticed that `get_token_transaction_hashes` returns a str. But this makes no sense. Can you please force it to return an evm transaction directly so that both here and in the one other place it's used we get a proper EvmTransaction?

What's more make sure to handle any deserialization errors (though if etherscan behaves okay there should be none)

File: rotkehlchen/chain/evm/transactions.py [hunk 15]
```diff
@@ -447,6 +460,71 @@ def _get_erc20_transfers_for_ranges(
...
+                    for log_entry in raw_receipt_data['logs']:
+                        if len(log_entry['topics']) == 0:
+                            continue
+
+                        topic_raw = log_entry['topics'][0]
+                        topic = hexstring_to_bytes(topic_raw)
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/transactions.py:489]: This can raise `DeserializationError` too. It needs to be handled.

File: rotkehlchen/chain/evm/transactions.py [hunk 16]
```diff
@@ -447,6 +460,71 @@ def _get_erc20_transfers_for_ranges(
...
+                        return False
+
+                    log.debug(f'Address detection: queried {self.evm_inquirer.chain_name} ERC20 Transfers for {address} -> range {start_ts} - {end_ts}')  # noqa: E501
+        except RemoteError as e:
+            self.msg_aggregator.add_error(
+                f'Got error "{e!s}" while querying {self.evm_inquirer.chain_name} '
+                f'token transactions from Etherscan. address: {address} spam detection failed',
+            )
+            return False
+        return True
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/transactions.py:526]:
```suggestion
            return False

        return True
```

File: rotkehlchen/tests/utils/polygon_pos.py (hunk 12 above)
Comment (LefterisJP) [reply rotkehlchen/tests/utils/polygon_pos.py:5]: No what I mean is what did you achieve with this change.Isn't it the same thing? I don't understand what you did here. You just put the endpoint out in a variable that's only used in this file. It makes no sense.

Comment (yabirgb) [reply rotkehlchen/tests/utils/polygon_pos.py:5]: It is used here https://github.com/rotki/rotki/pull/6285/files#diff-c9a958383afbcea23aa81d29d4ec9aebbdef01d05fddc3769d18fd3d0b9f2e8bR21

File: rotkehlchen/externalapis/etherscan.py [hunk 17]
```diff
@@ -451,17 +463,34 @@ def get_token_transaction_hashes(
...
+        hashes: set[tuple[EVMTxHash, Timestamp]] = set()
         while True:
             result = self._query(module='account', action='tokentx', options=options)
             last_ts = deserialize_timestamp(result[0]['timeStamp']) if len(result) != 0 else None
             for entry in result:
-                timestamp = deserialize_timestamp(entry['timeStamp'])
+                try:
+                    timestamp = deserialize_timestamp(entry['timeStamp'])
+                except DeserializationError as e:
+                    self.msg_aggregator.add_error(f"Failed to query information for transaction {entry['hash']}. Check the logs for more information")  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/externalapis/etherscan.py:474]: absolutely not. This will result in terrible spam. Please remove.

File: rotkehlchen/externalapis/etherscan.py [hunk 18]
```diff
@@ -451,17 +463,34 @@ def get_token_transaction_hashes(
...
+
                 if timestamp > last_ts and len(hashes) >= TRANSACTIONS_BATCH_NUM:  # type: ignore
                     yield _hashes_tuple_to_list(hashes)
                     hashes = set()
                     last_ts = timestamp
-                hashes.add((entry['hash'], timestamp))
+                try:
+                    hashes.add((deserialize_evm_tx_hash(entry['hash']), timestamp))
+                except DeserializationError as e:
+                    self.msg_aggregator.add_error(f"Failed to query information for transaction {entry['hash']}. Check the logs for more information")  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/externalapis/etherscan.py:488]: Same here. If we wanted to notify the user it should be a different type of websocket message that can be aggregated by the frontend.

**Review by LefterisJP — COMMENTED** [review]
Some more comments. Also in general please either be more focused or if it's not possible use an extension for typos. You make them way too often.

**Review by LefterisJP — APPROVED** [review]
LGTM. The failing tests are ... well .. fucking uniswap subgraph again
